app/scripts/modules/core/src/pipeline/config/stages/evaluateVariables/EvaluateVariablesStageConfig.tsx

```diff
@@ -1,48 +1,232 @@
+import { IStage } from 'core/domain';
+import { FormikStageConfig, IFormikStageConfigInjectedProps, IStageConfigProps } from 'core/pipeline';
+import {
+  FormValidator,
+  FormikFormField,
+  ILayoutProps,
+  IStageForSpelPreview,
+  IValidator,
+  LayoutContext,
+  LayoutProvider,
+  Markdown,
+  SpelInput,
+  StandardFieldLayout,
+  TextInput,
+  Tooltip,
+  ValidationMessage,
+  errorMessage,
+  useIsMountedRef,
+} from 'core/presentation';
+
+import { ExecutionAndStagePicker, IExecutionAndStagePickerProps } from './ExecutionAndStagePicker';
+
+import './EvaluateVariablesStageConfig.less';
 
 export interface IEvaluatedVariable {
   key: string;
   value: string;
 }
 
-export class EvaluateVariablesStageConfig extends React.Component<IStageConfigProps> {
-  private expand(variables: { [key: string]: string }): IEvaluatedVariable[] {
-    return map(variables, (value, key): IEvaluatedVariable => ({ key, value }));
-  }
-
-  private mapChanged = (key: string, values: { [key: string]: string }) => {
-    this.props.updateStageField({ [key]: this.expand(values) });
-  };
-
-  public render() {
-    const {
-      stage: { variables = [] },
-      pipeline,
-    } = this.props;
-
-    // Flattens an array of objects {key, value} into a single object with the respective keys/values
-    const variablesObject = variables.reduce(
-      // tslint:disable-next-line:prefer-object-spread
-      (acc: any, { key, value }: IEvaluatedVariable) => Object.assign(acc, { [key]: value }),
-      {},
-    );
-
-    return (
-      <div className="form-horizontal">
-        <StageConfigField label="Variables to evaluate">
-          <MapEditor
-            model={variablesObject}
-            allowEmpty={true}
-            onChange={(v: any) => this.mapChanged('variables', v)}
-            pipeline={pipeline}
-            valueCanContainSpel={true}
+const variableNameValidator: IValidator = (val: string, label: string) =>
+  !val.match(/^[a-zA-Z_][a-zA-Z0-9_]+$/) && errorMessage(`${label} should consist only of letters, numbers, or underscore`);
+
+const duplicateKeyValidatorFactory = (variables: IEvaluatedVariable[] = []) => {
+  const keyCounts = countBy(variables.map(x => x.key), x => x);
+  return (key: string) => keyCounts[key] > 1 && `Duplicate key '${key}'`;
+};
+
+export function validateEvaluateVariablesStage(stage: IStage) {
+  const formValidator = new FormValidator(stage);
+  const duplicateKeyValidator = duplicateKeyValidatorFactory(stage.variables);
+  formValidator.field('variables').withValidators(
+    formValidator.arrayForEach(item => {
+      item
+        .field('key', 'Variable Name')
+        .required()
+        .withValidators(variableNameValidator, duplicateKeyValidator);
+      item.field('value', 'Expression').required();
+    }),
+  );
+  return formValidator.validateForm();
+}
+
+function PreviewConfiguration(props: IExecutionAndStagePickerProps) {
+  return (
+    <div style={{ marginLeft: '16px', marginRight: '16px' }}>
```

Comment:
I'd like to see stuff like this become `sp-margin-l-xaxis` so we can keep spacing a bit less loosey goosey. You happened to pick the value we use for those classes in this case, but still :smile: